### DISCUSSION TITLE
Restrict account creation based on OIDC `groups` claim

### DISCUSSION THREAD

**Discussion by nick-pascucci-spire** [outline/outline#3296] [Ideas]
Hi,

I'm looking to set up Outline as an internal wiki system for my company. We need to restrict access to the system to a subset of our employees, so a purely domain-based approach won't work for us. Conveniently we have an OIDC group that we can use to give access. I'm thinking that a minimal approach would be to:

1. Add an environment variable, `OIDC_REQUIRE_GROUP`, which can be set to the group that must be claimed by the OIDC token on user account creation.
2. Extend the existing OIDC flow to extract the `groups` claim and check if the required group is present there.
3. If not, throw an auth error; otherwise, proceed.

This mirrors the approach for domain-based filtering. Is this something you would consider? I'm happy to write it up and submit a PR if so.

**Comment by tommoor** [comment]
This feels like quite a specific case that it would be better not to add yet another environment config for – they're getting a bit out of hand 😄 .  It would likely be better handled in the UI as part of OIDC -> Outline group mapping configuration, we're thinking that all external group mapping is likely to be an enterprise feature.

**Reply by ghi820** [reply]
Group mapping should not be considered "enterprise" it's absolute basics when using oidc providers. While self hosted OIDC providers (i.e. Keycloak) might give you the possibility of filtering OIDC really shines when using federated auth (e.g. via Github oAuth) only allowing certain groups (and possibly mapping them to the outline groups) should be considered the baseline for a somewhat complete OIDC implementation.

While trying to do all this in the Provider might be viable for some (those hosting the provider and also managing the users) it becomes more difficult the moment you want to have a more hands off approach.

For me I'm developing a student project, for which I've deployed outline on k8s fronted OIDC with Keycloak federating with Github. While implementing the groups in the fronting Keycloak this takes away from the flexibility (and is fairly hard to implement) as like most OIDC providers Keycloak "just provides authentication and metadata" and does not really implement authorization, which in the OIDC context should be implemented on the side of the application.

**Reply by almereyda** [reply]
Have you had a look at https://github.com/outline/outline/pull/2568 to check if it meets your requirements? For me the first three groups provided by OIDC upon first login/signup managed to get passed and mapped to Outline groups, but only in random order.

We could start off from there and see if this kind of mapping can be provided by the community, which often expresses a need for it, and thus gives back a little to the developers for their enterprise version.

**Reply by nick-pascucci-spire** [reply]
I'd be happy to proceed with a group mapping approach like in #2568. That PR seems to have languished a bit, but it's unclear if that's because it's unwanted or just paused.

**Reply by almereyda** [reply]
I feel @EternalDeiwos could not complete with updating the PR, since priorities possibly shifted meanwhile, and there was no actual feedback from the maintainers about if (1) this is in any way wanted and (2) would be accepted for merging. The primary requirement here could be to assume this fulfills all unit test, and is also tested in an integration test. For the amount of possible combinations with generic OpenID Connect providers #2205 this means a lot of maintenance that can come from this. I can understand why this would not be the primary objective of the maintainers.

Secondly, can we assume this discussion here to be a duplicate of #2206, and could therefore be closed?

In case this is still wanted, we could also organise as some kind of community initiative, and continue with bringing that branch into a mergeable state.

---

*Edit:* It's a bit a pity that linking to issues, pull requests or discussions here in the GitHub discussions does not create the backlinking headlines that we see between issues and pull requests. Eventually, if this is accepted as a legitimate claim, we could also promote #2206 into a proper issue, and work from there.
